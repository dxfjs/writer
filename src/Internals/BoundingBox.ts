import { point3d, point3d_t } from './TagsManager';

export type boundingBox_t = {
	topLeft: point3d_t;
	bottomRight: point3d_t;
};

export const createBoundingBox = (
	topLeft: point3d_t,
	bottomRight: point3d_t
): boundingBox_t => {
	return {
		topLeft,
		bottomRight,
	};
};

export default class BoundingBox {
	public static centerRadiusBBox(center: point3d_t, radius: number) {
		return createBoundingBox(
			point3d(center.x - radius, center.y + radius, 0),
			point3d(center.x + radius, center.y - radius, 0)
		);
	}

	public static pointBBox(point: point3d_t) {
		return createBoundingBox(
			point3d(point.x - 100, point.y + 100, 0),
			point3d(point.x + 100, point.y - 100, 0)
		);
	}

	public static lineBBox(startPoint: point3d_t, endPoint: point3d_t) {
		const maxX = Math.max(startPoint.x, endPoint.x);
		const minX = Math.min(startPoint.x, endPoint.x);
		const maxY = Math.max(startPoint.y, endPoint.y);
		const minY = Math.min(startPoint.y, endPoint.y);
		return createBoundingBox(
			point3d(minX, maxY, 0),
			point3d(maxX, minY, 0)
		);
	}

	public static verticesBBox(vertices: point3d_t[]) {
		const xCoordinates = vertices.map((vertex) => vertex.x);
		const yCoordinates = vertices.map((vertex) => vertex.y);
		const maxX = Math.max(...xCoordinates);
		const minX = Math.min(...xCoordinates);
		const maxY = Math.max(...yCoordinates);
		const minY = Math.min(...yCoordinates);
		return createBoundingBox(
			point3d(minX, maxY, 0),
			point3d(maxX, minY, 0)
		);
	}

	public static boundingBox(boundingBoxes: boundingBox_t[]): boundingBox_t {
		if (boundingBoxes.length === 0)
			return BoundingBox.pointBBox(point3d(0, 0, 0));
		return BoundingBox.verticesBBox(
			boundingBoxes.reduce((vertices: point3d_t[], bBox) => {
				return [...vertices, bBox.topLeft, bBox.bottomRight];
			}, [])
		);
	}

	public static boundingBoxCenter(boundingBox: boundingBox_t): point3d_t {
		const x =
			boundingBox.topLeft.x +
			(boundingBox.bottomRight.x - boundingBox.topLeft.x) / 2;
		const y =
			boundingBox.bottomRight.y +
			(boundingBox.topLeft.y - boundingBox.bottomRight.y) / 2;
		return point3d(x, y, 0);
	}

	public static boundingBoxHeight(boundingBox: boundingBox_t) {
		return boundingBox.topLeft.y - boundingBox.bottomRight.y;
	}
}
