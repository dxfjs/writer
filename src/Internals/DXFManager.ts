import Tag                          from "./Tag";
import DXFInterface                 from "./Interfaces/DXFInterface";
import NameComponent                from "./Components/NameComponent";
import ColorComponent               from "./Components/ColorComponent";
import LayerComponent               from "./Components/LayerComponent";
import PointComponent               from "./Components/PointComponent";
import HandleComponent              from "./Components/HandleComponent";
import LineTypeComponent            from "./Components/LineTypeComponent";
import StandardComponent            from "./Components/StandardComponent";
import TextStyleComponent           from "./Components/TextStyleComponent";
import ThicknessComponent           from "./Components/ThicknessComponent";
import TrueColorComponent           from "./Components/TrueColorComponent";
import EntityTypeComponent          from "./Components/EntityTypeComponent";
import SubclassMarkerComponent      from "./Components/SubclassMarkerComponent";
import SoftOwnerHandleComponent     from "./Components/SoftOwnerHandleComponent";
import HardOwnerHandleComponent     from "./Components/HardOwnerHandleComponent";
import SoftPointerHandleComponent   from "./Components/SoftPointerHandleComponent";
import HardPointerHandleComponent   from "./Components/HardPointerHandleComponent";

export default class DXFManager implements DXFInterface {
    get handle(): string {
        return this._handle;
    }
    set handle(value: string) {
        this._handle = value;
    }
    static versions = {
        R12:    "AC1009", R13:      "AC1012", R14:      "AC1014",
        R2000:  "AC1015", R2004:    "AC1018", R2007:    "AC1021",
        R2010:  "AC1024", R2013:    "AC1027"//, R2018:    "AC1032"
    };
    static colors = {
        Red:        1, Green:      3, Cyan:       4,
        Blue:       5, Magenta:    6, White:      7,
        Black:      0, Yellow:     2,
    }
    static units = {
        Unitless:           0,  Inches:         1,  Feet:           2,
        Miles:              3,  Millimeters:    4,  Centimeters:    5,
        Meters:             6,  Kilometers:     7,  Microinches:    8,
        Mils:               9,  Yards:          10, Angstroms:      11,
        Nanometers:         12, Microns:        13, Decimeters:     14,
        Decameters:         15, Hectometers:    16, Gigameters:     17,
        AstronomicalUnits:  18, LightYears:     19, Parsecs:        20,
        USSurveyFeet:       21, USSurveyInch:   22, USSurveyYard:   23,
        USSurveyMile:       24,
    }
    static version:         string = DXFManager.versions.R2007;
    static handleSeed:      number = 0;
    private _handle:        string;
    static currentLayer:    string = '0';
    public constructor() {
        this._handle = this.handleSeed();
    }

    public isInteger(value: number): boolean {
        return Number(value) === value && value % 1 === 0;
    }
    public isFloat(value: number): boolean {
        return Number(value) === value && value % 1 !== 0;
    }
    public isSupported(version: string): boolean {
        const fileVersion = parseInt(DXFManager.version.replace('AC', ''));
        const tagVersion = parseInt(version.replace('AC', ''));
        return tagVersion <= fileVersion;
    }
    public handleSeed (): string
    {
        DXFManager.handleSeed++;
        return DXFManager.handleSeed.toString(16).toUpperCase();
    }

    tags(): Tag[] {
        return [];
    }

    public stringify(): string {
        return this.tags().reduce((str, tag) => {
            if (this.isSupported(tag.version)) {

                return `${str}${tag.stringify()}`;
            } else {
                console.log(this.isSupported(tag.version));
            }
            return str;
        }, '');
    }

    public color(index: number): Tag[] {
        return new ColorComponent(index).tags();
    }
    public entityType(name: string): Tag[] {
        return new EntityTypeComponent(name).tags();
    }
    public hand(handle: string): Tag[] {
        return new HandleComponent(handle).tags();
    }
    public layer(layer: string): Tag[] {
        return new LayerComponent(layer).tags();
    }
    public lineType(lineType: string): Tag[] {
        return new LineTypeComponent(lineType).tags();
    }
    public name(name: string, groupCode: number = 2): Tag[] {
        return new NameComponent(name, groupCode).tags();
    }

    public point(
        x: number, y: number, z: number = 0,
        is3D: boolean = false, digit: number = 0
    ): Tag[] {
        return new PointComponent(x, y, z, is3D, digit).tags();
    }
    public standard(tagsArray: [number, number | string][]): Tag[] {
        return new StandardComponent(tagsArray).tags();
    }
    public subclassMarker(subclass: string): Tag[] {
        return new SubclassMarkerComponent(subclass).tags();
    }
    public textStyle(textStyleName: string): Tag[] {
        return new TextStyleComponent(textStyleName).tags();
    }
    public thickness(thickness : number): Tag[] {
        return new ThicknessComponent(thickness).tags();
    }
    public trueColor(
        red: number, green: number, blue: number, digit: number = 0
    ): Tag[] {
        return new TrueColorComponent(red, green, blue, digit).tags();
    }
    public softPointerHandle(softPointerHandle: string, digit: number = 0): Tag[] {
        return new SoftPointerHandleComponent(softPointerHandle, digit).tags();
    }
    public softOwnerHandle(softOwnerHandle: string, digit: number = 0): Tag[] {
        return new SoftOwnerHandleComponent(softOwnerHandle, digit).tags();
    }
    public hardPointerHandle(hardPointerHandle: string, digit: number = 0): Tag[] {
        return new HardPointerHandleComponent(hardPointerHandle, digit).tags();
    }
    public hardOwnerHandle(hardOwnerHandle: string, digit: number = 0): Tag[] {
        return new HardOwnerHandleComponent(hardOwnerHandle, digit).tags();
    }
};
