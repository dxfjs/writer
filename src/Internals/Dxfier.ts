import { vec2_t, vec3_t } from './Helpers'

export type tag_t = {
	code: number;
	value: number | string;
};

export class Dxfier {
  readonly lines: (string | number)[]
  constructor() {
    this.lines = []
  }

  push(code: number, value?: string | number) {
    if (value != null) this.lines.push(code, value)
  }

  stringify() {
    return this.lines.join('\u000A')
  }

  start(name: string) {
    this.push(0, 'SECTION')
    this.push(2, name)
  }

  end() {
    this.push(0, 'ENDSEC')
  }

  variableName(name: string) {
    this.push(9, name)
  }

  type(entityType: string) {
    this.push(0, entityType)
  }

  primaryText(primaryText: string) {
    this.push(1, primaryText)
  }

  name(name: string, code = 2) {
    this.push(code, name)
  }

  handle(handle: string) {
    this.push(5, handle)
  }

  lineType(lineType?: string) {
    this.push(6, lineType)
  }

  textStyle(textStyle: string) {
    this.push(7, textStyle)
  }

  layerName(layerName?: string) {
    this.push(8, layerName)
  }

  point2d(point?: vec2_t, digit = 0) {
    this.push(10 + digit, point?.x)
    this.push(20 + digit, point?.y)
  }

  point3d(point?: vec3_t, digit = 0) {
    this.point2d(point, digit)
    this.push(30 + digit, point?.z)
  }

  elevation(elevation?: number) {
    this.push(38, elevation)
  }

  thickness(thickness?: number) {
    this.push(39, thickness)
  }

  visibilty(visibilty?: boolean) {
    if (visibilty != null) this.push(60, visibilty ? 0 : 1)
  }

  colorNumber(colorNumber?: number) {
    this.push(62, colorNumber)
  }

  subclassMarker(subclassMarker?: string) {
    this.push(100, subclassMarker)
  }
}
