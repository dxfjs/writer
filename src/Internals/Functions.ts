import { DxfTag } from './Interfaces'

export function tag(code: number, value: number | string): DxfTag {
  return { code, value }
}

export function stringByteSize(value: string) {
  return new Blob([value]).size
}

export function stringChunksSplit(value: string, length = 255) {
  const chunks: string[] = []
  const tempChunk: string[] = []
  for (let i = 0; i < value.length; i++) {
    const char = value[i]
    if (tempChunk.length === length || i === value.length - 1) {
      chunks.push(tempChunk.join(''))
      tempChunk.length = 0
    } else tempChunk.push(char)
  }
  return chunks
}
