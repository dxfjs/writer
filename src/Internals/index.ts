export * from './BoundingBox'
export * from './Colors'
export * from './DefinedApplication'
export * from './Dxfier'
export * from './Enums'
export * from './ExtendedData'
export * from './Functions'
export * from './Handle'
export * from './HatchPatterns'
export * from './Helpers'
export * from './Interfaces'
export * from './TrueColor'
