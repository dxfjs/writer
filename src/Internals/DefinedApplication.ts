import { Dxfier, tag_t } from './Dxfier'
import { DxfInterface } from './Interfaces/DxfInterface'

export default class DxfDefinedApplication implements DxfInterface {
  readonly name: string
  readonly tags: tag_t[]

  constructor(name: string) {
    this.name = name
    this.tags = []
  }

  add(code: number, value: string) {
    this.tags.push({ code, value })
  }

  dxfy(dx: Dxfier) {
    dx.push(102, `{${this.name}`)
    for (const tag of this.tags) {
      dx.push(tag.code, tag.value)
    }
    dx.push(102, '}')
  }
}
