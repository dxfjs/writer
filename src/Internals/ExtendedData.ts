import { DxfInterface, DxfTag } from './Interfaces'
import { stringChunksSplit, tag } from './Functions'
import { Dxfier } from './Dxfier'
import { vec3_t } from './Helpers'

export class ExtendedData implements DxfInterface {
  name: string

  #tags: DxfTag[]

  constructor(name: string) {
    this.name = name
    this.#tags = []
  }

  clear() {
    this.#tags.length = 0
  }

  string(string: string) {
    stringChunksSplit(string).forEach((chunk) =>
      this.#tags.push(tag(1000, chunk))
    )
  }

  beginList() {
    this.#tags.push(tag(1002, '{'))
  }

  endList() {
    this.#tags.push(tag(1002, '}'))
  }

  layerName(name: string) {
    this.#tags.push(tag(1003, name))
  }

  binaryData(data: string) {
    stringChunksSplit(data).forEach((chunk) =>
      this.#tags.push(tag(1004, chunk))
    )
  }

  databaseHandle(handle: string) {
    this.#tags.push(tag(1005, handle))
  }

  point(point: vec3_t) {
    this.#tags.push(tag(1010, point.x))
    this.#tags.push(tag(1020, point.y))
    this.#tags.push(tag(1030, point.z))
  }

  position(position: vec3_t) {
    this.#tags.push(tag(1011, position.x))
    this.#tags.push(tag(1021, position.y))
    this.#tags.push(tag(1031, position.z))
  }

  displacement(displacement: vec3_t) {
    this.#tags.push(tag(1012, displacement.x))
    this.#tags.push(tag(1022, displacement.y))
    this.#tags.push(tag(1032, displacement.z))
  }

  direction(direction: vec3_t) {
    this.#tags.push(tag(1013, direction.x))
    this.#tags.push(tag(1023, direction.y))
    this.#tags.push(tag(1033, direction.z))
  }

  real(real: number) {
    this.#tags.push(tag(1040, real))
  }

  distance(distance: number) {
    this.#tags.push(tag(1041, distance))
  }

  scale(scale: number) {
    this.#tags.push(tag(1042, scale))
  }

  integer(integer: number) {
    this.#tags.push(tag(1070, integer))
  }

  long(long: number) {
    this.#tags.push(tag(1071, long))
  }

  dxfy(dx: Dxfier): void {
    dx.push(1001, this.name)
    dx.push(1002, '{')
    this.#tags.forEach((t) => dx.push(t.code, t.value))
    dx.push(1002, '}')
  }
}
