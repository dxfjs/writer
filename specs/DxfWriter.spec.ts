import {
  AttachmentPoint,
  DimStyleFlags,
  ImageDefResolutionUnits,
  LWPolylineFlags,
  LWPolylineVertex,
  TextLineSpacingStyle,
} from 'Sections'
import { Colors, Units, point2d, point3d } from 'Internals'
import { describe, expect, it } from 'vitest'
import { DxfWriter } from 'DxfWriter'
import { writeFileSync } from 'fs'

describe('DxfWriter', () => {
  it('should cover all code source', () => {
    const dxf = new DxfWriter()
    dxf.addPoint(0, 0, 0)
    const vertices: LWPolylineVertex[] = [
      {
        point: point2d(0, 0),
        startingWidth: 5,
        endWidth: 2,
      },
      {
        point: point2d(100, 100),
      },
      {
        point: point2d(100, 200),
      },
      {
        point: point2d(0, 300),
      },
    ]

    dxf.addLWPolyline(vertices, {
      flags: LWPolylineFlags.Closed,
    })
    dxf.addLine(point3d(0, 0, 0), point3d(100, 100, 0))

    dxf.addLType('AXES', '____ _ ', [4, -1, 1, -1])

    dxf.addRectangle(point2d(400, 400), point2d(600, 200), {
      elevation: 30,
      constantWidth: 10,
      lineType: 'AXES',
      lineTypeScale: 10,
      visible: true,
    })

    dxf.addRectangle(point2d(400, 400), point2d(600, 200), {
      fillet: 50,
      visible: false,
    })

    const circleBlock = dxf.addBlock('circle')
    circleBlock.addCircle(point3d(0, 0, 0), 50)
    circleBlock.addRectangle(
      point2d(-35.3553, 35.3553),
      point2d(35.3553, -35.3553)
    )

    dxf.addInsert(circleBlock.name, point3d(0, 0, 0))

    const controlPoints = [
      point3d(0, 0, 0),
      point3d(10, 10, 0),
      point3d(20, 10, 0),
      point3d(30, 20, 0),
    ]

    dxf.addSpline({
      controlPoints,
      fitPoints: controlPoints,
    })

    dxf.addArc(point3d(0, 0, 0), 10, 0, 45)

    dxf.addEllipse(
      point3d(100, 100, 0),
      point3d(50, 0, 0),
      0.5,
      0,
      2 * Math.PI
    )

    const face = dxf.add3dFace(
      point3d(0, 0, 50),
      point3d(0, 100, 50),
      point3d(100, 100, 50),
      point3d(100, 0, 50)
    )
    face.setEdgesVisible(true)
    face.setFirstEdgeVisible(false)
    face.setSecondEdgeVisible(true)
    face.setSecondEdgeVisible(false)
    face.setThirdEdgeVisible(false)
    face.setFourthEdgeVisible(false)
    face.setEdgesVisible(false)

    dxf.addCircle(point3d(0, 0, 0), 50)

    dxf.addPolyline3D([
      { point: point3d(0, 0, 0) },
      { point: point3d(50, 40, 0) },
      { point: point3d(70, 63, 2) },
    ])

    dxf.addImage(
      '.\\test - image.png',
      'test - image',
      point3d(700, 600, 0),
      1792,
      1280,
      433.54,
      360 - 359.74
    )

    const id = dxf.document.objects.addImageDef('jhdjshfsd')
    id.resolutionUnits = ImageDefResolutionUnits.Inch

    dxf.document.objects.addEntryToRoot('test', 'FF')

    dxf.tables.addView({
      name: 'testview',
      backClipping: 0,
      frontClipping: 0,
      isUCSAssociated: false,
      lensLength: 50.0,
      renderMode: 0,
      targetPoint: point3d(0, 0, 0),
      twistAngle: 0,
      viewCenter: point2d(40.36, 15.86),
      viewDirection: point3d(0, 0, 1),
      viewHeight: 17.91,
      viewMode: 0,
      viewWidth: 20.01,
    })

    const v = dxf.tables.addView({
      name: 'view2',
      backClipping: 0,
      frontClipping: 0,
      isUCSAssociated: true,
      lensLength: 50.0,
      renderMode: 0,
      targetPoint: point3d(0, 0, 0),
      twistAngle: 0,
      viewCenter: point2d(40.36, 15.86),
      viewDirection: point3d(0, 0, 1),
      viewHeight: 17.91,
      viewMode: 0,
      viewWidth: 20.01,
      liveSectionObjectHandle: 'dd',
      visualStyleObjectHandle: 'ff',
      backgroundObjectHandle: 'AA',
    })
    v.isCameraPlottable = true

    dxf.addText(point3d(0, 0, 0), 5, 'test')

    const test = dxf.addLayer('test', Colors.Blue)
    const newTest = dxf.addLayer('test', Colors.Red)

    dxf.addLayer('ggg', Colors.Black, 'not-exist')
    const e = dxf.tables.layerTable.exist('ggg')
    const ne = dxf.tables.layerTable.exist('not-exist')

    dxf.header
    dxf.entities
    dxf.currentLayer
    dxf.units
    dxf.layer('ggg')
    dxf.setZeroLayerAsCurrent()
    dxf.setVariable('$DIMTXT', { 40: 10 })
    const dim = dxf.addDimStyle('test')
    dim.flags = DimStyleFlags.None
    dxf.addAlignedDim(point3d(0, 0), point3d(50, 50))
    dxf.addAlignedDim(point3d(0, 0), point3d(50, 50), {
      offset: 10,
      insertionPoint: point3d(0, 0),
      rotation: 'auto',
      attachmentPoint: AttachmentPoint.MiddleCenter,
      textLineSpacingStyle: TextLineSpacingStyle.AtLeast,
    })
    dxf.addDiameterDim(point3d(0, 0), point3d(50, 50))
    dxf.addDiameterDim(point3d(0, 0), point3d(50, 50), { leaderLength: 40 })
    dxf.addRadialDim(point3d(0, 0), point3d(50, 50), { leaderLength: 400 })
    dxf.addLinearDim(point3d(0, 0), point3d(50, 50))
    dxf.addLinearDim(point3d(0, 0), point3d(50, 50), {
      offset: 10,
      insertionPoint: point3d(0, 0),
      angle: 30,
      linearType: 1,
      rotation: 'auto',
      middlePoint: point3d(0, 0),
    })
    dxf.addLinearDim(point3d(0, 0), point3d(50, 50), {
      offset: 10,
      insertionPoint: point3d(0, 0),
      linearType: 1,
      rotation: 'auto',
    })
    dxf.document.entities.modelSpace.addAngularPointsDim(
      point3d(0, 0),
      point3d(50, 50),
      point3d(50, 0)
    )

    const dline = {
      start: point3d(0, 0),
      end: point3d(50, 50),
    }
    dxf.addAngularLinesDim(dline, dline, point3d(50, 50))

    dxf.setCurrentLayerName('not-exist-exist')
    dxf.setUnits(Units.Unitless)

    dxf.tables.addUcs('ucs')
    dxf.tables.addAppId('testest')

    dxf.addLType('test', '---', [1, 1, 1])
    dxf.addLType('test', '---', [3, 3, 3])

    const _str = dxf.stringify()
    writeFileSync('examples/tests.dxf', _str)

    expect(e).toBe(true)
    expect(ne).toBe(false)
    expect(newTest).toBe(test)
    expect(dxf.stringify()).toMatchSnapshot()
  })
})
