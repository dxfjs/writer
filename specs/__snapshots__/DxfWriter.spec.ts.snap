// Vitest Snapshot v1

exports[`DxfWriter > should cover all code source 1`] = `
"0
SECTION
2
HEADER
9
$ACADVER
1
AC1021
9
$LASTSAVEDBY
1
@tarikjabiri/dxf
9
$HANDSEED
5
3E
9
$INSUNITS
70
0
9
$CLAYER
8
not-exist-exist
9
$DIMTXT
40
10
9
$VIEWCTR
10
672.945209627366
20
599.9999999999999
0
ENDSEC
0
SECTION
2
CLASSES
0
ENDSEC
0
SECTION
2
TABLES
0
TABLE
2
VPORT
5
1
330
0
100
AcDbSymbolTable
70
1
0
VPORT
5
19
330
1
100
AcDbSymbolTableRecord
100
AcDbViewportTableRecord
2
*Active
70
0
10
0
20
0
11
1
21
1
12
672.945209627366
22
599.9999999999999
13
0
23
0
14
10
24
10
15
10
25
10
16
0
26
0
36
1
17
0
27
0
37
0
40
1491.780838509464
41
2
42
50
43
0
44
0
50
0
51
0
71
0
72
100
73
1
74
3
75
0
76
1
77
0
78
0
281
0
65
1
110
0
120
0
130
0
111
1
121
0
131
0
112
0
122
1
132
0
79
0
146
0
348
10020
60
7
61
5
292
1
282
1
141
0
142
0
63
250
421
3358443
0
ENDTAB
0
TABLE
2
LTYPE
5
2
330
0
100
AcDbSymbolTable
70
4
0
LTYPE
5
12
330
2
100
AcDbSymbolTableRecord
100
AcDbLinetypeTableRecord
2
ByBlock
70
0
3

72
65
73
0
40
0
0
LTYPE
5
13
330
2
100
AcDbSymbolTableRecord
100
AcDbLinetypeTableRecord
2
ByLayer
70
0
3

72
65
73
0
40
0
0
LTYPE
5
14
330
2
100
AcDbSymbolTableRecord
100
AcDbLinetypeTableRecord
2
Continuous
70
0
3
Solid line
72
65
73
0
40
0
0
LTYPE
5
1D
330
2
100
AcDbSymbolTableRecord
100
AcDbLinetypeTableRecord
2
AXES
70
0
3
____ _ 
72
65
73
4
40
7
49
4
74
0
49
-1
74
0
49
1
74
0
49
-1
74
0
0
ENDTAB
0
TABLE
2
LAYER
5
3
330
0
100
AcDbSymbolTable
70
3
0
LAYER
5
15
330
3
100
AcDbSymbolTableRecord
100
AcDbLayerTableRecord
2
0
70
0
62
7
6
Continuous
370
0
390
0
0
LAYER
5
36
330
3
100
AcDbSymbolTableRecord
100
AcDbLayerTableRecord
2
test
70
0
62
5
6
Continuous
370
0
390
0
0
LAYER
5
37
330
3
100
AcDbSymbolTableRecord
100
AcDbLayerTableRecord
2
ggg
70
0
62
0
6
Continuous
370
0
390
0
0
ENDTAB
0
TABLE
2
STYLE
5
4
330
0
100
AcDbSymbolTable
70
1
0
STYLE
5
16
330
4
100
AcDbSymbolTableRecord
100
AcDbTextStyleTableRecord
2
Standard
70
0
40
0
41
1
50
0
71
0
42
1
3
txt
4

0
ENDTAB
0
TABLE
2
VIEW
5
5
330
0
100
AcDbSymbolTable
70
1
0
VIEW
5
34
330
5
100
AcDbSymbolTableRecord
100
AcDbViewTableRecord
2
testview
70
0
40
17.91
10
40.36
20
15.86
41
20.01
11
0
21
0
31
1
12
0
22
0
32
0
42
50
43
0
44
0
50
0
71
0
281
0
72
0
0
ENDTAB
0
TABLE
2
UCS
5
6
330
0
100
AcDbSymbolTable
70
0
0
ENDTAB
0
TABLE
2
APPID
5
7
330
0
100
AcDbSymbolTable
70
1
0
APPID
5
17
330
7
100
AcDbSymbolTableRecord
100
AcDbRegAppTableRecord
2
ACAD
70
0
0
ENDTAB
0
TABLE
2
DIMSTYLE
5
8
330
0
100
AcDbSymbolTable
70
2
100
AcDbDimStyleTable
0
DIMSTYLE
105
18
330
8
100
AcDbSymbolTableRecord
100
AcDbDimStyleTableRecord
2
Standard
70
0
340
16
0
DIMSTYLE
105
38
330
8
100
AcDbSymbolTableRecord
100
AcDbDimStyleTableRecord
2
test
70
0
0
ENDTAB
0
TABLE
2
BLOCK_RECORD
5
9
330
0
100
AcDbSymbolTable
70
3
0
BLOCK_RECORD
5
C
330
9
100
AcDbSymbolTableRecord
100
AcDbBlockTableRecord
2
*Model_Space
70
0
280
1
281
0
0
BLOCK_RECORD
5
F
330
9
100
AcDbSymbolTableRecord
100
AcDbBlockTableRecord
2
*Paper_Space
70
0
280
1
281
0
0
BLOCK_RECORD
5
20
330
9
100
AcDbSymbolTableRecord
100
AcDbBlockTableRecord
2
circle
70
0
280
1
281
0
0
ENDTAB
0
ENDSEC
0
SECTION
2
BLOCKS
0
BLOCK
5
D
330
C
100
AcDbEntity
8
not-exist-exist
100
AcDbBlockBegin
2
*Model_Space
70
0
10
0
20
0
30
0
3
*Model_Space
1

0
ENDBLK
5
E
100
AcDbEntity
8
0
100
AcDbBlockEnd
0
BLOCK
5
10
330
F
100
AcDbEntity
8
0
100
AcDbBlockBegin
2
*Paper_Space
70
0
10
0
20
0
30
0
3
*Paper_Space
1

0
ENDBLK
5
11
100
AcDbEntity
8
0
100
AcDbBlockEnd
0
BLOCK
5
21
330
20
100
AcDbEntity
8
0
100
AcDbBlockBegin
2
circle
70
0
10
0
20
0
30
0
3
circle
1

0
CIRCLE
5
23
330
21
100
AcDbEntity
8
0
100
AcDbCircle
10
0
20
0
30
0
40
50
0
LWPOLYLINE
5
24
330
21
100
AcDbEntity
8
0
100
AcDbPolyline
90
4
70
1
43
0
38
0
39
0
10
-35.3553
20
35.3553
10
35.3553
20
35.3553
10
35.3553
20
-35.3553
10
-35.3553
20
-35.3553
0
ENDBLK
5
22
100
AcDbEntity
8
0
100
AcDbBlockEnd
0
ENDSEC
0
SECTION
2
ENTITIES
0
POINT
5
1A
330
D
100
AcDbEntity
8
0
100
AcDbPoint
10
0
20
0
30
0
0
LWPOLYLINE
5
1B
330
D
100
AcDbEntity
8
0
100
AcDbPolyline
90
4
70
1
38
0
39
0
10
0
20
0
40
5
41
2
10
100
20
100
10
100
20
200
10
0
20
300
0
LINE
5
1C
330
D
100
AcDbEntity
8
0
100
AcDbLine
10
0
20
0
30
0
11
100
21
100
31
0
0
LWPOLYLINE
5
1E
330
D
100
AcDbEntity
8
0
6
AXES
48
10
60
0
100
AcDbPolyline
90
4
70
1
43
10
38
30
39
0
10
400
20
400
10
600
20
400
10
600
20
200
10
400
20
200
0
LWPOLYLINE
5
1F
330
D
100
AcDbEntity
8
0
60
1
100
AcDbPolyline
90
8
70
1
43
0
38
0
39
0
10
400
20
350
42
-0.414213562373095
10
450
20
400
10
550
20
400
42
-0.414213562373095
10
600
20
350
10
600
20
250
42
-0.414213562373095
10
550
20
200
10
450
20
200
42
-0.414213562373095
10
400
20
250
0
INSERT
5
25
330
D
100
AcDbEntity
8
0
100
AcDbBlockReference
2
circle
10
0
20
0
30
0
41
1
42
1
43
1
50
0
70
1
71
1
44
0
45
0
0
SPLINE
5
26
330
D
100
AcDbEntity
8
0
100
AcDbSpline
70
8
71
3
72
8
73
4
74
4
42
0.0000001
43
0.0000001
42
0.0000000001
40
0
40
0
40
0
40
0
40
1
40
1
40
1
40
1
41
1
10
0
20
0
30
0
10
10
20
10
30
0
10
20
20
10
30
0
10
30
20
20
30
0
11
0
21
0
31
0
11
10
21
10
31
0
11
20
21
10
31
0
11
30
21
20
31
0
0
ARC
5
27
330
D
100
AcDbEntity
8
0
100
AcDbCircle
10
0
20
0
30
0
40
10
100
AcDbArc
50
0
51
45
0
ELLIPSE
5
28
330
D
100
AcDbEntity
8
0
100
AcDbEllipse
10
100
20
100
30
0
11
50
21
0
31
0
40
0.5
41
0
42
6.283185307179586
0
3DFACE
5
29
330
D
100
AcDbEntity
8
0
100
AcDbFace
10
0
20
0
30
50
11
0
21
100
31
50
12
100
22
100
32
50
13
100
23
0
33
50
70
15
0
CIRCLE
5
2A
330
D
100
AcDbEntity
8
0
100
AcDbCircle
10
0
20
0
30
0
40
50
0
POLYLINE
5
2B
330
D
100
AcDbEntity
8
0
100
AcDb3dPolyline
66
1
10
0
20
0
30
0
39
0
70
0
40
0
41
0
71
0
72
0
73
0
74
0
75
0
0
VERTEX
5
2D
100
AcDbEntity
8
0
100
AcDbVertex
100
AcDb3dPolylineVertex
10
0
20
0
30
0
70
32
0
VERTEX
5
2E
100
AcDbEntity
8
0
100
AcDbVertex
100
AcDb3dPolylineVertex
10
50
20
40
30
0
70
32
0
VERTEX
5
2F
100
AcDbEntity
8
0
100
AcDbVertex
100
AcDb3dPolylineVertex
10
70
20
63
30
2
70
32
0
SEQEND
5
2C
100
AcDbEntity
8
0
0
IMAGE
5
31
330
D
100
AcDbEntity
8
0
100
AcDbRasterImage
90
0
10
700
20
600
30
0
11
0.24192831263950734
21
-0.00109784339408965
31
0
12
0.00109784339408965
22
0.24192831263950734
32
0
13
1792
23
1280
340
30
70
3
280
1
281
50
282
50
283
0
360
32
71
1
91
2
14
-0.5
24
-0.5
14
1791.5
24
1279.5
290
1
0
TEXT
5
35
330
D
100
AcDbEntity
8
0
100
AcDbText
10
0
20
0
30
0
40
5
1
test
7
STANDARD
100
AcDbText
0
DIMENSION
5
39
330
D
100
AcDbEntity
8
0
100
AcDbDimension
70
1
71
5
100
AcDbAlignedDimension
13
0
23
0
33
0
14
50
24
50
34
0
0
DIMENSION
5
3A
330
D
100
AcDbEntity
8
0
100
AcDbDimension
10
50
20
50
30
0
70
3
71
5
100
AcDbDiametricDimension
15
0
25
0
35
0
0
DIMENSION
5
3B
330
D
100
AcDbEntity
8
0
100
AcDbDimension
10
50
20
50
30
0
70
4
71
5
100
AcDbRadialDimension
15
0
25
0
35
0
0
DIMENSION
5
3C
330
D
100
AcDbEntity
8
0
100
AcDbDimension
70
0
71
5
100
AcDbAlignedDimension
13
0
23
0
33
0
14
50
24
50
34
0
50
0
100
AcDbRotatedDimension
0
DIMENSION
5
3D
330
D
100
AcDbEntity
8
0
100
AcDbDimension
10
50
20
50
30
0
70
2
71
5
100
AcDb2LineAngularDimension
13
0
23
0
33
0
14
50
24
50
34
0
15
0
25
0
35
0
16
50
26
50
36
0
0
ENDSEC
0
SECTION
2
OBJECTS
0
DICTIONARY
5
A
330
0
100
AcDbDictionary
281
1
3
ACAD_GROUP
350
B
3
ACAD_IMAGE_DICT
350
33
0
DICTIONARY
5
B
330
A
100
AcDbDictionary
281
0
0
IMAGEDEF
5
30
330
33
102
{ACAD_REACTORS
330
33
330
32
102
}
100
AcDbRasterImageDef
1
.\\\\test - image.png
10
1792
20
1280
11
1
21
1
280
1
281
0
0
IMAGEDEF_REACTOR
5
32
330
0
100
AcDbRasterImageDefReactor
90
2
330
31
0
DICTIONARY
5
33
330
A
100
AcDbDictionary
281
0
3
test - image
350
30
0
ENDSEC
0
EOF"
`;
